File: 01_GOF23/DP_examples/Go/01_Creational/Singleton/singleton.go
package singleton

import (
	"fmt"
	"sync"
)

// 非线程安全单例
type Singleton struct {
	Value int
}

var instance *Singleton

func GetSingleton() *Singleton {
	if instance == nil {
		instance = &Singleton{Value: 0}
	}
	return instance
}

// 线程安全单例
type SingletonLock struct {
	Value int
}

var (
	lockInstance *SingletonLock
	lock         sync.Mutex
)

func GetSingletonLock() *SingletonLock {
	if lockInstance == nil {
		lock.Lock()
		defer lock.Unlock()
		if lockInstance == nil {
			lockInstance = &SingletonLock{Value: 0}
		}
	}
	return lockInstance
}

// 使用 sync.Once 实现线程安全单例
type SingletonOnce struct {
	Value int
}

var (
	onceInstance *SingletonOnce
	once         sync.Once
)

func GetSingletonOnce() *SingletonOnce {
	once.Do(func() {
		onceInstance = &SingletonOnce{Value: 0}
	})
	return onceInstance
}

// 内联单例
type SingletonInline struct {
	Value int
}

var inlineInstance = &SingletonInline{Value: 0}

func GetSingletonInline() *SingletonInline {
	return inlineInstance
}

// 继承式单例
type SingletonInherit interface {
	Log()
}

type MonoSingleton struct {
	Value int
}

var monoInstance *MonoSingleton

func GetMonoSingleton() *MonoSingleton {
	if monoInstance == nil {
		monoInstance = &MonoSingleton{Value: 0}
	}
	return monoInstance
}

func (m *MonoSingleton) Log() {
	fmt.Println("MonoSingleton is created")
}

type GameSingleton struct {
	Value int
}

var gameInstance *GameSingleton

func GetGameSingleton() *GameSingleton {
	if gameInstance == nil {
		gameInstance = &GameSingleton{Value: 0}
	}
	return gameInstance
}

func (g *GameSingleton) Log() {
	fmt.Println("GameSingleton is created")
}

File: 01_GOF23/DP_examples/Lua/01_Creational/Singleton/Singleton.lua
-- 常规单例
Singleton = {}
Singleton.__index = Singleton

local instance = nil

function Singleton.getInstance()
    if instance == nil then
        instance = setmetatable({}, Singleton)
    end
    return instance
end

function Singleton:showMessage()
    print("Singleton instance")
end

-- 继承单例基类
Singleton_Inherit = {}
Singleton_Inherit.__index = Singleton_Inherit
function Singleton_Inherit:log() end

MonoSingleton = setmetatable({}, { __index = Singleton_Inherit })
MonoSingleton.__index = MonoSingleton
local monoInstance = nil
function MonoSingleton.getInstance()
    if monoInstance == nil then
        monoInstance = setmetatable({}, MonoSingleton)
    end
    return monoInstance
end
function MonoSingleton:log()
    print("MonoSingleton is created")
end

GameSingleton = setmetatable({}, { __index = Singleton_Inherit })
GameSingleton.__index = GameSingleton
local gameInstance = nil
function GameSingleton.getInstance()
    if gameInstance == nil then
        gameInstance = setmetatable({}, GameSingleton)
    end
    return gameInstance
end
function GameSingleton:log()
    print("GameSingleton is created")
end